File: app/backend/schemas/ceim.py
"""
Canonical Emergency Incident Model (CEIM).

A versioned, provenance-tracked intermediate representation of one incident,
per the mentor's 2026-09-02 direction: the AI never generates a
protocol-specific format directly. It populates this internal model, and
protocol-specific adapters (NG112 today, EDXL-SitRep planned) translate FROM
it on demand. See ResQKit_Canonical_Incident_Model.md for the full design
rationale and the EDXL-SitRep/NG112 background.

Every meaningful fact is wrapped in `Fact`, carrying WHERE it came from and
HOW confident we are — the thing today's IncidentState/brief.ts has none of.

Safety invariant (enforced in services/ceim.py, not just here): a victim's
`responsive`/`breathing` facts must always carry source="button_selected".
The AI is never allowed to originate or overwrite those two fields, because
they are what routes routeProcedure() to CPR. CEIM is never itself a source
of medical guidance — knowledge.ts's curated PROCEDURES remain the only
guidance shown to the user.
"""

from datetime import datetime
from typing import Generic, List, Literal, Optional, TypeVar

from pydantic import BaseModel, Field

CEIM_SCHEMA_VERSION = "0.1.0"

T = TypeVar("T")

CeimSource = Literal[
    "bystander_stated",  # free text from the AI interview, folded in verbatim
    "button_selected",  # fixed-choice triage/hazards/kit UI - always highest confidence
    "device_sensor",  # GPS fix etc.
    "ai_inferred",  # model-extracted/derived from bystander_stated text
    "not_recorded",
]

CeimConfidence = Literal["high", "medium", "low"]


class Fact(BaseModel, Generic[T]):
    """One piece of information, with where it came from and how sure we are."""

    value: Optional[T] = Field(default=None, description="The fact's value, or null if unknown.")
    source: CeimSource = Field(default="not_recorded", description="Where this fact came from.")
    confidence: CeimConfidence = Field(default="low", description="How confident we are in this fact.")
    note: Optional[str] = Field(
        default=None, description="Short justification, mainly used for ai_inferred facts."
    )


class CeimLocation(BaseModel):
    """Scene location, mixing device-sensor facts with bystander description."""

    latitude: Optional[Fact[float]] = None
    longitude: Optional[Fact[float]] = None
    accuracy_m: Optional[Fact[float]] = None
    description: Optional[Fact[str]] = Field(
        default=None, description="Free-text location note (landmark, address detail)."
    )


class CeimVictim(BaseModel):
    """
    One victim's state.

    `responsive`/`breathing` MUST always carry source="button_selected" -
    see the module docstring's safety invariant. Everything else may carry
    richer bystander-stated or ai_inferred detail.
    """

    index: int = Field(..., description="0-based victim index, matching the app's completedSteps ordering.")
    responsive: Optional[Fact[str]] = None
    breathing: Optional[Fact[str]] = None
    age_band: Optional[Fact[str]] = None
    injury_type: Optional[Fact[str]] = None
    trapped: Optional[Fact[str]] = None
    condition_description: Optional[Fact[str]] = Field(
        default=None, description="Free-text detail beyond the fixed triage fields."
    )


class CeimHazard(BaseModel):
    """One observed hazard, matched against the app's hazard whitelist where possible."""

    code: Optional[Fact[str]] = Field(
        default=None, description="Whitelist hazard code if the description matched one."
    )
    description: Fact[str] = Field(..., description="Raw hazard text, always present.")


class CeimIncident(BaseModel):
    """The full canonical model for one incident."""

    ceim_schema_version: str = Field(default=CEIM_SCHEMA_VERSION)
    generated_at: datetime = Field(..., description="When this report was generated.")
    content_pack_version: str = Field(..., description="Content pack version active when generated.")
    incident_type: Optional[Fact[str]] = None
    called_112: Optional[Fact[str]] = None
    location: CeimLocation = Field(default_factory=CeimLocation)
    victim_count: Optional[Fact[int]] = None
    victims: List[CeimVictim] = Field(default_factory=list)
    hazards: List[CeimHazard] = Field(default_factory=list)
    kit_items: List[Fact[str]] = Field(
        default_factory=list, description="Equipment confirmed on scene, button_selected from the kit stage."
    )
    scene_observations: List[Fact[str]] = Field(
        default_factory=list, description="Free-text narrative snippets, one per interview prompt."
    )
    additional_notes: Optional[Fact[str]] = None
    degraded: bool = Field(
        default=False, description="True if AI extraction failed and this is skeleton-only."
    )


class KnownFacts(BaseModel):
    """
    Already-captured, button/sensor-sourced incident fields (never AI-touched).

    `hazards`/`kit_items` are optional because the interview stage runs
    BEFORE the hazards/kit stages in the wizard - a first report generation
    naturally has neither. A "Regenerate report" action, fired once those
    stages are done, resends this same request with them populated.
    """

    incident_type: Optional[str] = None
    called_112: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    accuracy_m: Optional[float] = None
    location_note: Optional[str] = None
    victim_count: Optional[int] = None
    responsive: Optional[str] = None
    breathing: Optional[str] = None
    injury: Optional[str] = None
    age_band: Optional[str] = None
    trapped: Optional[str] = None
    hazards: List[str] = Field(default_factory=list, description="Confirmed hazard whitelist codes.")
    kit_items: List[str] = Field(default_factory=list, description="Confirmed kit item whitelist codes.")


class InterviewAnswerIn(BaseModel):
    """One answered (or skipped) interview prompt."""

    prompt_id: str = Field(..., description="Stable id matching INTERVIEW_PROMPTS on the client.")
    prompt_text: str = Field(..., description="The prompt text shown to the user, for model context.")
    answer_text: str = Field(default="", description="The bystander's free-text answer, if any.")


class CeimGenerateRequest(BaseModel):
    """Everything needed to generate one CEIM report."""

    known_facts: KnownFacts = Field(..., description="Already-captured button/sensor facts.")
    interview_answers: List[InterviewAnswerIn] = Field(
        default_factory=list, description="Free-text interview answers, in prompt order."
    )
    content_pack_version: str = Field(default="unknown")


class CeimGenerateResponse(BaseModel):
    """Always 200 - see CeimService.generate for the graceful-degrade contract."""

    ceim: CeimIncident
    degraded: bool = Field(..., description="True if the model call failed or produced no extraction.")
    model: str = Field(..., description="Model used, or 'none' if no interview text was provided.")

File: app/mobile/src/lib/ceim.ts
/**
 * Canonical Emergency Incident Model (CEIM) — mobile mirror.
 *
 * Field-for-field mirror of app/backend/schemas/ceim.py. Kept in sync by
 * hand for now (no shared workspace package yet, same accepted gap already
 * documented for knowledge.ts/brief.ts and the i18n locale files) — treat
 * any drift between the two as a defect. See ResQKit_Canonical_Incident_Model.md
 * for the full design rationale.
 *
 * Safety invariant (enforced server-side in services/ceim.py, mirrored here
 * only as documentation): a victim's `responsive`/`breathing` facts always
 * carry source='button_selected'. The AI interview below never populates or
 * overwrites those two fields, under any circumstance — they are what
 * routes routeProcedure() to CPR.
 */
import { IncidentState } from './storage';

export const CEIM_SCHEMA_VERSION = '0.1.0';

export type CeimSource =
  | 'bystander_stated'
  | 'button_selected'
  | 'device_sensor'
  | 'ai_inferred'
  | 'not_recorded';

export type CeimConfidence = 'high' | 'medium' | 'low';

export interface Fact<T> {
  value: T | null;
  source: CeimSource;
  confidence: CeimConfidence;
  note?: string | null;
}

export interface CeimLocation {
  latitude?: Fact<number> | null;
  longitude?: Fact<number> | null;
  accuracy_m?: Fact<number> | null;
  description?: Fact<string> | null;
}

export interface CeimVictim {
  index: number;
  responsive?: Fact<string> | null;
  breathing?: Fact<string> | null;
  age_band?: Fact<string> | null;
  injury_type?: Fact<string> | null;
  trapped?: Fact<string> | null;
  condition_description?: Fact<string> | null;
}

export interface CeimHazard {
  code?: Fact<string> | null;
  description: Fact<string>;
}

export interface CeimIncident {
  ceim_schema_version: string;
  generated_at: string;
  content_pack_version: string;
  incident_type?: Fact<string> | null;
  called_112?: Fact<string> | null;
  location: CeimLocation;
  victim_count?: Fact<number> | null;
  victims: CeimVictim[];
  hazards: CeimHazard[];
  scene_observations: Fact<string>[];
  additional_notes?: Fact<string> | null;
  degraded: boolean;
}

/* ------------------------------------------------------------------ *
 * Request/response contract for POST /api/v1/resqkit/ceim/generate
 * ------------------------------------------------------------------ */

export interface KnownFacts {
  incident_type: string | null;
  called_112: string | null;
  latitude: number | null;
  longitude: number | null;
  accuracy_m: number | null;
  location_note: string | null;
  victim_count: number | null;
  responsive: string | null;
  breathing: string | null;
  injury: string | null;
  age_band: string | null;
  trapped: string | null;
}

export interface InterviewAnswerIn {
  prompt_id: string;
  prompt_text: string;
  answer_text: string;
}

export interface CeimGenerateRequest {
  known_facts: KnownFacts;
  interview_answers: InterviewAnswerIn[];
  content_pack_version: string;
}

export interface CeimGenerateResponse {
  ceim: CeimIncident;
  degraded: boolean;
  model: string;
}

/**
 * Derives the request's `known_facts` from an incident's already-captured
 * button/sensor fields, so no screen has to hand-assemble this shape ad hoc.
 */
export function buildKnownFactsFromIncident(incident: IncidentState): KnownFacts {
  return {
    incident_type: incident.context,
    called_112: incident.called112,
    latitude: incident.latitude,
    longitude: incident.longitude,
    accuracy_m: incident.accuracy,
    location_note: incident.locationNote || null,
    victim_count: incident.victimCount,
    responsive: incident.responsive || null,
    breathing: incident.breathing || null,
    injury: incident.injury || null,
    age_band: incident.ageBand || null,
    trapped: incident.trapped || null,
  };
}

/**
 * The fixed prompt set (design decision A2): pre-written, not model-generated,
 * so answering them costs zero AI latency. Every prompt is optional — a
 * bystander must never be blocked by a mandatory field. Ordered to build on
 * what triage already captured, per the interview's placement right after it.
 */
export const INTERVIEW_PROMPTS: { id: string; prompt: string }[] = [
  {
    id: 'scene_description',
    prompt: 'In your own words — what happened, and what do you see right now?',
  },
  {
    id: 'victim_condition',
    prompt:
      'Describe how the injured person looks and acts right now, beyond what you already selected (skin colour, bleeding amount, pain, movement, speech).',
  },
  {
    id: 'hazards_observed',
    prompt:
      'What hazards or dangers do you see, even small ones (traffic, fire, spills, unstable objects, water, crowd)?',
  },
  {
    id: 'access_detail',
    prompt: 'Is anyone trapped or hard to reach? Is anything blocking access for you or for responders?',
  },
  {
    id: 'anything_else',
    prompt:
      'Anything else responders should know before they arrive — weather, number of vehicles, people helping, anything unusual?',
  },
];
